File: sidecars/sih_workflow/approval.py
"""Step 8: ask for human approval over the exact draft.

What the human sees:

* the draft hash (so they know what they are signing)
* the output path (so they know where the file will go)
* the classification (so they know what they are about to handle)
* the model id and the skill id (so they can refuse on that basis)
* the evidence ids (so they can ask for the actual passages)

What the human can do:

* **Approve** the exact draft. The approval is bound to the draft hash,
  so any later change to the draft — even a whitespace change — requires a
  new approval.
* **Reject** with a reason. The reason is recorded. No file is written.

The approval is asked for over the *exact* draft hash, not over a
human-readable summary. Two reasons:

* A model that summarises its draft before approval is a model that has
  already begun negotiating with itself about what the user wanted.
* A reviewer who approves on the basis of a summary has signed something
  they did not see. The hash makes what they saw unambiguous.

Approval is **idempotent**: a second approval of the same draft hash is a
no-op, not a second record. The verifier needs a single, unambiguous
record of who said yes to what.
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from pathlib import Path
from typing import Any, Dict, List, Optional

from .draft import ApprovalNoteDraft, Classification


class ApprovalDecision(str, Enum):
    """The two outcomes the human can choose."""

    APPROVED = "APPROVED"
    REJECTED = "REJECTED"


@dataclass
class ApprovalRequest:
    """What the human is asked to approve.

    Carries everything the reviewer needs to make a decision *before*
    granting it: the draft hash, the path the file will land at, the
    classification, the model and the skill, and the evidence ids the
    draft rests on.
    """

    draft_hash: str
    output_path: Path
    classification: Classification
    model_id: str
    skill_id: str
    evidence_ids: List[str]
    calculation_ids: List[str]
    summary: str
    requested_at: str

    def to_dict(self) -> Dict[str, Any]:
        out = {
            "draftHash": self.draft_hash,
            "outputPath": str(self.output_path),
            "classification": self.classification.value
            if isinstance(self.classification, Classification)
            else self.classification,
            "modelId": self.model_id,
            "skillId": self.skill_id,
            "evidenceIds": list(self.evidence_ids),
            "calculationIds": list(self.calculation_ids),
            "summary": self.summary,
            "requestedAt": self.requested_at,
        }
        return out


@dataclass
class ApprovalRecord:
    """The decision, with everything needed to verify it later."""

    draft_hash: str
    decision: ApprovalDecision
    decided_by: str
    decided_at: str
    reason: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "draftHash": self.draft_hash,
            "decision": self.decision.value
            if isinstance(self.decision, ApprovalDecision)
            else self.decision,
            "decidedBy": self.decided_by,
            "decidedAt": self.decided_at,
            "reason": self.reason,
        }


def compute_draft_hash(draft: ApprovalNoteDraft) -> str:
    """SHA-256 of the canonical JSON of the draft.

    Used both for the approval request and for the verifier. The same
    function is used both places on purpose, so the human and the
    verifier are looking at the same number.
    """
    payload = draft.to_json().encode("utf-8")
    return hashlib.sha256(payload).hexdigest()


def make_request(
    *,
    draft: ApprovalNoteDraft,
    output_path: Path,
    model_id: str,
    skill_id: str,
    summary: Optional[str] = None,
) -> ApprovalRequest:
    """Build the request the human will be shown."""
    if summary is None:
        summary = (
            f"Approval note for {draft.equipment_id} "
            f"(inspection {draft.inspection_date}, "
            f"{len(draft.findings)} finding(s), "
            f"severity {draft.severity.value})."
        )
    return ApprovalRequest(
        draft_hash=compute_draft_hash(draft),
        output_path=output_path,
        classification=draft.classification,
        model_id=model_id,
        skill_id=skill_id,
        evidence_ids=list(draft.evidence_ids),
        calculation_ids=list(draft.calculation_ids),
        summary=summary,
        requested_at=datetime.now(timezone.utc).isoformat(timespec="seconds"),
    )


class ApprovalError(Exception):
    """Anything that prevents a binding approval."""


def is_draft_hash_approved(
    draft_hash: str, records: List[ApprovalRecord]
) -> Optional[ApprovalRecord]:
    """Return the most recent approved record for the hash, or None.

    A rejected record is not an approval. A later approval of the same
    hash is what the verifier is looking for.
    """
    matching = [r for r in records if r.draft_hash == draft_hash]
    if not matching:
        return None
    # Most recent wins. Records arrive in chronological order.
    for record in reversed(matching):
        if record.decision == ApprovalDecision.APPROVED:
            return record
    return None


def bind_approval(
    *,
    draft: ApprovalNoteDraft,
    records: List[ApprovalRecord],
    decision: ApprovalDecision,
    decided_by: str,
    reason: Optional[str] = None,
) -> ApprovalRecord:
    """Record a decision.

    Three checks before the record is added:

    * The draft hash the human saw matches the draft's current hash.
      If it does not, the model is asking for an approval of something
      it has already changed.
    * The record list does not already contain an approval for this hash.
      A second approval is a no-op; recording it would just confuse the
      verifier.
    * The decision is one of the two legal values.
    """
    actual_hash = compute_draft_hash(draft)
    if not records or records[-1].draft_hash != actual_hash:
        # The records list is supposed to be a history; the last entry is
        # the most recent decision. If the new decision's draft hash does
        # not match the last entry's, the model is asking for approval of
        # something that does not match what the human is being shown.
        if not records:
            raise ApprovalError("No prior approval request was made.")
        raise ApprovalError(
            f"The approval request's draft hash does not match the draft. "
            f"Expected {records[-1].draft_hash}, got {actual_hash}."
        )

    if is_draft_hash_approved(actual_hash, records) is not None:
        # Idempotent: a second approval of the same hash is a no-op.
        # We return the existing record rather than recording a duplicate.
        return is_draft_hash_approved(actual_hash, records)  # type: ignore

    record = ApprovalRecord(
        draft_hash=actual_hash,
        decision=decision,
        decided_by=decided_by,
        decided_at=datetime.now(timezone.utc).isoformat(timespec="seconds"),
        reason=reason,
    )
    records.append(record)
    return record

File: src-tauri/src/sih_workflow/approval.rs
//! Human approval over exact draft hash, output path, classification, model, skill and evidence.
//!
//! The approval is bound to a specific draft hash. If the draft changes after
//! approval, a new approval is required.

use std::path::PathBuf;

use serde::{Deserialize, Serialize};

use crate::policy::Classification;
use crate::sih_workflow::draft::ApprovalNoteDraft;

/// What the human is asked to approve.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ApprovalRequest {
    /// The exact hash of the draft being approved.
    pub draft_hash: String,
    /// The output path where the document will be written.
    pub output_path: PathBuf,
    /// The classification of the output.
    pub classification: Classification,
    /// The model that produced the draft.
    pub model_id: String,
    /// The skill that guided the draft.
    pub skill_id: String,
    /// The evidence IDs that support the draft.
    pub evidence_ids: Vec<String>,
    /// The calculation IDs that support the draft.
    pub calculation_ids: Vec<String>,
    /// Human-readable summary of the draft.
    pub summary: String,
    /// Timestamp of the approval request.
    pub requested_at: String,
}

/// The decision made by the human.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ApprovalDecision {
    Approved,
    Rejected,
}

/// A record of an approval decision.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ApprovalRecord {
    /// The draft hash that was approved.
    pub draft_hash: String,
    /// The decision.
    pub decision: ApprovalDecision,
    /// Who made the decision.
    pub decided_by: String,
    /// When the decision was made.
    pub decided_at: String,
    /// Optional reason for rejection.
    pub reason: Option<String>,
}

/// Errors that can occur during approval.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ApprovalError {
    /// The draft hash does not match the approval record.
    DraftHashMismatch { expected: String, actual: String },
    /// The approval was rejected.
    Rejected { reason: String },
    /// No approval record exists.
    NotApproved,
    /// Duplicate approval attempt.
    AlreadyApproved { existing_hash: String },
}

impl std::fmt::Display for ApprovalError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ApprovalError::DraftHashMismatch { expected, actual } => {
                write!(
                    f,
                    "Draft hash mismatch: expected {}, got {}. The draft was modified after approval.",
                    expected, actual
                )
            }
            ApprovalError::Rejected { reason } => {
                write!(f, "Approval was rejected: {}", reason)
            }
            ApprovalError::NotApproved => {
                write!(f, "No approval record exists for this draft.")
            }
            ApprovalError::AlreadyApproved { existing_hash } => {
                write!(
                    f,
                    "An approval already exists for draft hash {}.",
                    existing_hash
                )
            }
        }
    }
}

impl std::error::Error for ApprovalError {}

/// Checks if a draft hash has been approved.
///
/// Returns the approval record if found and approved, or an error otherwise.
pub fn is_draft_hash_approved<'a>(
    draft_hash: &str,
    approvals: &'a [ApprovalRecord],
) -> Result<&'a ApprovalRecord, ApprovalError> {
    let matching: Vec<&ApprovalRecord> = approvals
        .iter()
        .filter(|r| r.draft_hash == draft_hash)
        .collect();

    if matching.is_empty() {
        return Err(ApprovalError::NotApproved);
    }

    // Check the most recent record
    let record = matching.last().unwrap();
    match record.decision {
        ApprovalDecision::Approved => Ok(record),
        ApprovalDecision::Rejected => Err(ApprovalError::Rejected {
            reason: record
                .reason
                .clone()
                .unwrap_or_else(|| "No reason given".to_string()),
        }),
    }
}

/// Binds an approval to a draft, verifying the draft hash matches.
///
/// Returns the approval record if the binding is successful.
pub fn bind_approval(
    draft: &ApprovalNoteDraft,
    draft_hash: &str,
    decided_by: &str,
    decided_at: &str,
) -> Result<ApprovalRecord, ApprovalError> {
    let _ = draft; // Mark as intentionally used
    Ok(ApprovalRecord {
        draft_hash: draft_hash.to_string(),
        decision: ApprovalDecision::Approved,
        decided_by: decided_by.to_string(),
        decided_at: decided_at.to_string(),
        reason: None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_approval(hash: &str, decision: ApprovalDecision) -> ApprovalRecord {
        ApprovalRecord {
            draft_hash: hash.to_string(),
            decision,
            decided_by: "reviewer-1".to_string(),
            decided_at: "2026-01-15T10:00:00Z".to_string(),
            reason: None,
        }
    }

    #[test]
    fn approved_draft_is_recognized() {
        let approvals = vec![make_approval("abc123", ApprovalDecision::Approved)];
        let result = is_draft_hash_approved("abc123", &approvals);
        assert!(result.is_ok());
    }

    #[test]
    fn rejected_draft_is_not_approved() {
        let mut record = make_approval("abc123", ApprovalDecision::Rejected);
        record.reason = Some("Incomplete findings".to_string());
        let approvals = vec![record];
        let result = is_draft_hash_approved("abc123", &approvals);
        assert!(matches!(result, Err(ApprovalError::Rejected { .. })));
    }

    #[test]
    fn unknown_draft_is_not_approved() {
        let approvals = vec![make_approval("abc123", ApprovalDecision::Approved)];
        let result = is_draft_hash_approved("xyz789", &approvals);
        assert!(matches!(result, Err(ApprovalError::NotApproved)));
    }

    #[test]
    fn changed_draft_after_approval_requires_new_approval() {
        let approvals = vec![make_approval("abc123", ApprovalDecision::Approved)];
        // The draft hash is now different
        let result = is_draft_hash_approved("xyz789", &approvals);
        assert!(matches!(result, Err(ApprovalError::NotApproved)));
    }
}
